### PR TITLE
Py-side EFP and Mol parsing

### PR DESCRIPTION
**Pull request by loriab**
## Description
This gets the scfitertopy (scf iterations py-side, pb221, pcm py-side) fully operational again with efp py-side. Also extends pyvib2's Molecule dict (often labeled "molrec") init and validation to input parsing. This requires two extensions to molrec – one for efp and one for zmat/variables in geometry. Replaces all the c-side constructors (i.e., create_molecule_from_string, init_with_xyz), replacing them with a unified dict, str, arrays constructor.

This branch is presently master (reviewed) + pyvib2 (unreviewed) + scfitertopy (semi-reviewed) + newstuff, all rebased so it doesn't align with anything. After pyvib2 makes it into master, I'll rebase and separate into scfitertopy and newstuff, so it can be readily reviewed. This is just getting the commits posted so I can break psi locally yet again.

## Todos
Notable points that this PR has either accomplished or will accomplish.
* **Developer Interest**
  - [ ] Feature1
* **User-Facing for Release Notes**
  - [ ] Feature2

## Questions
- [ ] Question1

## Status
- [ ] Ready to go

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -143,6 +143,8 @@ install:
 - conda info -a
 - conda create -q -n p4env python=$PYTHON_VER numpy cmake ci-psi4-lt psi4-lt-mp gau2grid dftd3 gcp snsmp2 scipy -c psi4/label/dev -c psi4
 - source activate p4env
+- conda install hungarian networkx -c psi4
+- conda install deepdiff -c conda-forge
```
Comment (dgasmith) [inline .travis.yml]: Are these actively needed by Psi4? Seems like quite a few new depends.

File: psi4/driver/driver.py [hunk 2]
```diff
@@ -1274,12 +1324,22 @@ def hessian(name, **kwargs):
...
             dertype = 1
 
+    # At stationary point?
+    G0 = gradient(lowername, molecule=molecule, **kwargs)
+    translations_projection_sound, rotations_projection_sound = _energy_is_invariant(G0)
+    core.print_out('\n  Based on options and gradient (rms={:.2E}), recommend {}projecting translations and {}projecting rotations.\n'.
+                   format(G0.rms(), '' if translations_projection_sound else 'not ',
+                   '' if rotations_projection_sound else 'not '))
+    if not core.has_option_changed('FINDIF', 'FD_PROJECT'):
+        core.set_local_option('FINDIF', 'FD_PROJECT', rotations_projection_sound)
```
Comment (dgasmith) [inline psi4/driver/driver.py]: Cool!

File: psi4/driver/json_wrapper.py [hunk 3]
```diff
@@ -29,15 +29,15 @@
 Runs a JSON input psi file.
 """
 
+import os
+import copy
+import uuid
+
+import psi4
 from psi4.driver import driver
```
Comment (dgasmith) [inline psi4/driver/json_wrapper.py]: Is this reordering from PyCharm?

File: psi4/driver/molutil.py [hunk 4]
```diff
@@ -209,70 +58,115 @@ def molecule_get_attr(self, name):
...
+                         molecular_charge=None,
+                         molecular_multiplicity=None,
+
+                         missing_enabled_return='error',
+                         tooclose=0.1,
+                         zero_ghost_fragments=False,
+                         nonphysical=False,
+                         mtol=1.e-3,
+                         verbose=1,
+
```
Comment (dgasmith) [inline psi4/driver/molutil.py]: This is quite a few new arrays. Anything that needs to go back to the QC Schema?

File: psi4/driver/molutil.py [hunk 5]
```diff
@@ -305,12 +210,22 @@ def geometry(geom, name="default"):
...
+                                       enable_qm=True,
+                                       missing_enabled_return_qm='minimal',
+                                       enable_efp=True,
+                                       missing_enabled_return_efp='none')
+
+    molecule = core.Molecule.from_dict(molrec['qm'])
     molecule.set_name(name)
 
+    if 'efp' in molrec:
+        import pylibefp
```
Comment (dgasmith) [inline psi4/driver/molutil.py]: Does this need an import guard?

File: psi4/driver/qcdb/util/misc.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+
+def distance_matrix(a, b):
+    """Euclidean distance matrix between rows of arrays `a` and `b`. Equivalent to
+    `scipy.spatial.distance.cdist(a, b, 'euclidean')`. Returns a.shape[0] x b.shape[0] array.
+
+    """
+    assert a.shape[1] == b.shape[1], """Inner dimensions do not match"""
+    distm = np.zeros([a.shape[0], b.shape[0]])
+    for i in range(a.shape[0]):
```
Comment (dgasmith) [inline psi4/driver/qcdb/util/misc.py]: Maybe `distm[i] = np.linalg.norm(a[i] - b, axis=1)` to get a little broadcasting in?

File: psi4/driver/json_wrapper.py (hunk 3 above)
Comment (loriab) [reply psi4/driver/json_wrapper.py]: No, I just went through and cleared out a lot of unused imports and reorganized into standard modules, external modules, and psi4 relative imports. Then reordered by length of module name. The last is admittedly arbitrary, but imports were already jumbled so it seemed harmless and orderly.

File: psi4/driver/molutil.py (hunk 4 above)
Comment (loriab) [reply psi4/driver/molutil.py]: except for `name`, they're all in there. the last block (missing_ ... mtol) are controlling _how_ it's processed, so not for the schema.

File: .travis.yml (hunk 1 above)
Comment (loriab) [reply .travis.yml]: Yeah, pains me a bit, too. They're all there for the comparison of Mol to its dict repr. They're definitely needed for a few tests. Right now they're needed for everything b/c I put in an extra every-time check to satisfy myself it was working in volume. Been ok for a couple months now, so possibly removeable.

The deepdiff doesn't bother me - its conda and pip-able. but the hungarian and networkx are rather overkill for everyday dependencies. Will have to consider them further.

**Comment by loriab** [comment]
supplanted by #953

**Pull request closed without merging**